### PR TITLE
Port to container-based Travis infrastructure and fix code coverage collection

### PR DESCRIPTION
**Pull request by JoshRosen**
This patch updates the build to use Travis's container-based infrastructure and updates the code coverage configurations to fix Scala-based code coverage collection and enable branch coverage and Codecov reports for the Python tests.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,25 +1,33 @@
 language: scala
+cache:
+  directories:
+    - $HOME/.ivy2
 scala:
    - 2.10.4
-before_install:
-  - sudo pip install --upgrade conda py4j
```
Comment (JoshRosen) [inline .travis.yml]: Note that I explicitly do not install Py4J anymore, since forcing an upgrade seemed to break things.

File: python/run-tests [hunk 2]
```diff
@@ -1,6 +1,20 @@
...
 if [ -z "$SPARK_HOME" ]; then
    echo "$(tput setaf 1)Error: SPARK_HOME is not set, may not be able to run tests.$(tput sgr0)"
 fi
-PYTHON_PATH=./ nosetests --logging-level=INFO --detailed-errors --verbosity=2 --with-coverage --cover-html-dir=./htmlcov --cover-package=sparktestingbase --traverse-namespace 
+PYTHON_PATH=./ nosetests \
+    --logging-level=INFO \
+    --detailed-errors \
+    --verbosity=2 \
+    --with-coverage \
+    --cover-branches \
```
Comment (JoshRosen) [inline python/run-tests]: This was a nice excuse to play around with Codecov's new support for Python branch coverage.

File: build.sbt [hunk 3]
```diff
@@ -20,6 +20,11 @@ sparkComponents ++= Seq("core", "streaming", "sql", "hive", "streaming-kafka", "
 
 parallelExecution in Test := false
 
+coverageHighlighting := {
+  if (scalaBinaryVersion.value == "2.10") false
```
Comment (JoshRosen) [inline build.sbt]: As you noticed, Scoverage hits weird AST processing issues under Scala 2.10. These are fixed in 2.11, where we're able to re-enable coverage highlighting.

File: .travis.yml [hunk 4]
```diff
@@ -1,25 +1,33 @@
 language: scala
+cache:
+  directories:
+    - $HOME/.ivy2
```
Comment (JoshRosen) [inline .travis.yml]: The build is _waaaay_ faster the second time around when the cache kicks in. ~2 minutes roundtrip on Travis for me.

**Comment by JoshRosen** [comment]
By the way, now that Codecov is working you should check out its configuration options to see if you want it to comment and leave coverage diff comments on PRs: https://codecov.io/github/holdenk/spark-testing-base/features/comment

**Comment by holdenk** [comment]
Thanks so much! Having codecoverage working should be fun :)

File: .travis.yml (hunk 4 above)
Comment (holdenk) [inline .travis.yml]: Cool, I was wondering how to get the most of the caching with the new container based travis infra for open source projects. I'll give this a shot in some other projects too.

**Comment by holdenk** [comment]
rebased & merged into master, thanks! :)

**Pull request closed without merging**